ttnn/cpp/ttnn/operations/data_movement/copy/device/copy_program_factory.cpp

```diff
@@ -0,0 +1,226 @@
+            .set_page_size(output_cb_index, aligned_output_unit_size);
+        auto cb_output = tt::tt_metal::CreateCircularBuffer(program, all_cores, output_cb_config);
+    }
+
+    auto src_buffer = input.buffer();
+    auto dst_buffer = output.buffer();
+    bool src_is_dram = src_buffer->buffer_type() == tt::tt_metal::BufferType::DRAM ? 1 : 0;
+    bool dst_is_dram = dst_buffer->buffer_type() == tt::tt_metal::BufferType::DRAM ? 1 : 0;
+
+    std::vector<uint32_t> reader_compile_time_args, writer_compile_time_args;
+    if (tilized) {
+        reader_compile_time_args = {(uint32_t)src_is_dram};
+        writer_compile_time_args = {
+            (std::uint32_t) output_cb_index,
+            (std::uint32_t) dst_is_dram
+        };
+    } else {
+        bool src_stick_size_is_power_of_two = is_power_of_two_at_least_32(input_unit_size);
+        uint32_t src_log2_stick_size = src_stick_size_is_power_of_two ? (std::uint32_t)log2(input_unit_size) : 0;
+        reader_compile_time_args = {
+            (std::uint32_t) src0_cb_index,
+            (std::uint32_t) src_is_dram,
+            (std::uint32_t) src_stick_size_is_power_of_two,
+            (std::uint32_t) src_log2_stick_size
+        };
+        bool dst_stick_size_is_power_of_two = is_power_of_two_at_least_32(output_unit_size);
+        uint32_t dst_log2_stick_size = dst_stick_size_is_power_of_two ? (std::uint32_t)log2(output_unit_size) : 0;
+        writer_compile_time_args = {
+            (std::uint32_t) output_cb_index,
+            (std::uint32_t) dst_is_dram,
+            (std::uint32_t) dst_stick_size_is_power_of_two,
+            (std::uint32_t) dst_log2_stick_size
+        };
+    }
+    std::map<string, string> kernel_defines;
+    if (backwards) {
+        kernel_defines["BACKWARDS"] = "1";
+    }
+    tt::tt_metal::KernelHandle unary_reader_kernel_id = tt::tt_metal::CreateKernel(
+        program,
+        tilized ? "ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/dataflow/reader_unary_interleaved_start_id.cpp" : "ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/reader_unary_stick_layout_interleaved_start_id.cpp",
+        all_cores,
+        tt::tt_metal::ReaderDataMovementConfig(reader_compile_time_args, kernel_defines));
+
+    tt::tt_metal::KernelHandle unary_writer_kernel_id = tt::tt_metal::CreateKernel(
+        program,
+        tilized ? "ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/dataflow/writer_unary_interleaved_start_id.cpp" : "ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/writer_unary_stick_layout_interleaved_start_id.cpp",
+        all_cores,
+        tt::tt_metal::WriterDataMovementConfig(writer_compile_time_args, kernel_defines));
+
+    if (convert_dtype) {
+        vector<uint32_t> compute_kernel_args_group_1 = {
+            num_units_per_core_group_1
+        };
+        auto eltwise_unary_kernel_group_1 = tt::tt_metal::CreateKernel(
+            program,
+            "ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/compute/eltwise_copy.cpp",
```

Comment:
at some point we have to move this to ttnn

Reply:
any idea why is this in that location and not inside device folder of copy op?

Reply:
It's used in other OPs too like reshape and interleave_to_sharded. We should probably figure out a place to put all these kernels that are reused